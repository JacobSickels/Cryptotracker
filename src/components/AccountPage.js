import React from 'react';
import { connect } from 'react-redux';
import { Input } from 'react-materialize';

import { startEditCurrency } from '../actions/filters';

export class AccountPage extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            currency: props.currency
        }

    }

    onInputChange = (e) => {
        const currency = this.props.exchanges.find((element) => element.id === e.target.value);
        this.setState({ currency });
        this.props.startEditCurrency(currency);
    }

    render() {
        return (
            <div className="page-container">
                <div className="container">
                    <h2>My Account Page</h2>

                    <h1>Set Base Currency</h1>
                    <Input s={5} type='select' label="Select Currency" value={this.state.currency.id} onChange={this.onInputChange}>
                        {
                            this.props.exchanges.map((currency) =>
                                <option key={currency.id} value={currency.id}>{currency.name}</option>
                            )
                        }
                    </Input>
                </div>
            </div>
        );
    }

}

const mapDispatchToProps = (dispatch) => {
    return {
        startEditCurrency: (currency) => dispatch(startEditCurrency(currency))
    };
};

const mapStateToProps = (state, props) => {
    return {
        exchanges: state.exchange,
        currency: state.filters.currency
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(AccountPage);