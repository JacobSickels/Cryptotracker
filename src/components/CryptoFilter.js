import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';

import { setStartDate, setEndDate } from '../actions/filters';

export class CryptoFilter extends React.Component {

    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    render() {
        return (
            <div className="date-container">
                <div className="container">
                    <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        minimumNights={0}
                        isOutsideRange={(day) => {
                            const today = moment();
                            return day >= today;
                        }}
                    />
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(CryptoFilter);