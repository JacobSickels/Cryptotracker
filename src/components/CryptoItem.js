import React from 'react';
import { Link } from 'react-router-dom'; //Eventually linking to the specific cryptos page for more specific filtering
import moment from 'moment';
import { connect } from 'react-redux';
import { ResponsiveContainer, LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend } from 'recharts';
import selectCryptos from '../selectors/cryptos';

export class CryptoItem extends React.Component {
    
    constructor(props) {
        super(props);

        this.state = {
            showTrendLine: true
        }
    }

    render() {
        return (
            <div className="chart">
                <div className="chart-title">
                    <div className="container">
                        <div className="chart-title--container">
                            <div>
                                <Link className="chart-title__link" to={`/filter/${this.props.name.toLowerCase()}`} >
                                    <h1> {this.props.name} </h1>
                                </Link>
                            </div>
                            <div className="chart-title__amount">
                                <h1>
                                    {this.props.currency_symbol} {this.props.current.amount}
                                </h1>
                            </div>
                            <div className="chart-title--percentage">
                                 <h1 style={{
                                     color:
                                        (this.props.percentage >= 0) ? 'green' : 'red'
                                     }}>
                                    {this.props.percentage.toFixed(2)}% 
                                </h1>
                            </div>
                        </div>
                    </div>
                </div>
                <div className="container">
                    <div className="button-container">
                        <button className="button--trend"
                            onClick={() => {
                                const show = this.state.showTrendLine;
                                this.setState({showTrendLine: !show})
                            }}>
                        {
                            (this.state.showTrendLine) ? 'Hide Trend' : 'Show Trend'
                        }      
                        </button>
                    </div>
                    <div className="chart-container">
                        <ResponsiveContainer height='100%' width='100%'>
                            <LineChart data={this.props.data} margin={{top: 5, right: 10, left: 30, bottom: 5}}>
                                <XAxis dataKey="timestamp"/>
                                <YAxis tickSize={10} dataKey="amount" domain={["dataMin", "dataMax"]} allowDataOverflow={true} />
                                <CartesianGrid strokeDasharray="3 3" stroke="#FFDDED"/>
                                <Tooltip/>
                                <Legend />
                                {
                                    (this.state.showTrendLine) && (
                                        <Line dot={false} type="monotone" dataKey="trend" stroke="#E477D4" strokeWidth={4}/>
                                    )
                                }
                                <Line dot={false} type="monotone" dataKey="amount" stroke="#7F4276"/>
                            </LineChart>
                        </ResponsiveContainer>
                    </div>
                </div>
            </div>
        );
    };
}

const convertToExchange = (exchangeRate, cryptos) => {
    const converted = cryptos.map((entry) => ({
        amount: (parseFloat(entry.amount) * exchangeRate).toFixed(2),
        timestamp: entry.timestamp
    }));

    return converted;
}

const getTrendData = (crypto) => {
    let x = [];
    
    for(let i = 0; i < crypto.length; i++) {
        x.push(i);
    }

    const xSum = x.map(num => num).reduce((prev, next) => prev + next);
    const ySum = crypto.map(entry => parseFloat(entry.amount)).reduce((prev, next) => prev + next);
    const averageX = xSum / crypto.length;
    const averageY = ySum / crypto.length;

    const topSlopeArray = [];
    for(let i = 0; i < crypto.length; i++) {
        topSlopeArray.push((x[i] - averageX) * (crypto[i].amount - averageY));
    }
    const topSlope = topSlopeArray.reduce((prev, next) => prev + next);
    const bottomSlope = x.map(num => (num - averageX) * (num - averageX)).reduce ((prev, next) => prev + next);

    const slope = topSlope / bottomSlope;
    const yIntercept = averageY - (slope * averageX);
    const percentage = (crypto[crypto.length - 1].amount - crypto[0].amount) / crypto[0].amount * 100;

    const data = crypto.map((entry, index) => ({
                timestamp: moment(entry.timestamp).format('ha'), 
                amount: parseFloat(entry.amount), 
                trend: slope * index + yIntercept
            }));

    return {
        data,
        percentage,
        current: crypto[crypto.length - 1]
    };
};

const mapStateToProps = (state, props) => {

    let crypto = selectCryptos(state.cryptos, state.filters)[props.name.toLowerCase()];

    const exchangeRate = parseFloat(state.filters.currency.exchange_rate);

    //converting amounts to exchange rate
    crypto = convertToExchange(exchangeRate, crypto);

    return {
        ...getTrendData(crypto),
        currency_symbol: state.filters.currency.symbol
    }
    
};

export default connect(mapStateToProps)(CryptoItem);