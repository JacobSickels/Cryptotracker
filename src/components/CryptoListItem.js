import React from 'react';
import { Link } from 'react-router-dom'; //Eventually linking to the specific cryptos page for more specific filtering
import moment from 'moment';
import momentTimezone from 'moment-timezone';


const CryptoListItem = (props) => (
    <div>
    {
        props.crypto.map(
            (entry) => <p key={entry.timestamp}> Amount: { entry.amount } Time: { moment(entry.timestamp).tz('America/Chicago').format() } </p>
        ) 
    }
    </div>
);
export default CryptoListItem;