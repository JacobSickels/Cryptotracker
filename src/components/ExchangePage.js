import React from 'react';
import { connect } from 'react-redux';
import { Row, Input, Col, Icon } from 'react-materialize';
import { startSetExchange } from '../actions/exchange';

import LoadingPage from './LoadingPage';

/*

need to load from and to elements into default state.


*/

export class ExchangePage extends React.Component {
    constructor(props) {
        super(props);
        const from = (this.props.from_element) ? this.props.from_element : null;
        const to = (this.props.to_element) ? this.props.to_element : null;
        this.state = {
            componentMounted: false,
            amount: 1,
            from_element: from,
            to_element: to,
            conversion: (amount) => {
                if(this.state.from_element === null) {
                    this.setState({ from_element: this.props.from_element });
                    this.setState({ to_element: this.props.to_element });
                }
                let conv = (this.state.from_element) ? (amount / this.state.from_element.exchange_rate) * this.state.to_element.exchange_rate : 0;
                return conv.toFixed(2);
            }

        }

        this.props.startSetExchange();
    }

    onToInputChange = (e) => {
        const target = this.props.exchanges.find((element) => element.id === e.target.value);
        console.log(e.target.value);
        this.setState({to_element: target});
    }

    onFromInputChange = (e) => {
        const target = this.props.exchanges.find((element) => element.id === e.target.value);
        console.log(e.target.value);
        this.setState({from_element: target});
    }

    onAmountChange = (e) => {
        let amount = e.target.value;
        if(!amount || isNaN(amount)) {
            amount = 1;
        }
        this.setState({ amount });
    }
    
    componentDidMount() {
        this.setState({componentMounted: true});
        console.log('mounted');
        console.log(this.props);
    }
    
    render() {
        
        if(!this.state.componentMounted || this.state.from_element === null || this.state.conversion(this.state.amount) === 0) {
            console.log('mounted', this.state.componentMounted);
            console.log('from', this.state.from_element);
            console.log('');
            return <LoadingPage />
        }

        return (
            <div className="page-container">
                <div className="exchange-display">
                    <div className="container"> 
                        <h1> {this.state.amount} {this.state.from_element.name}</h1>
                        <h1> {this.state.conversion(this.state.amount)} {this.state.to_element.name}</h1>
                    </div>
                </div>
                <div className="container exchange">
                    <Row className="selector">
                        <Input s={5} label="Input Amount" defaultValue='1' onChange={this.onAmountChange} />
                    </Row>
                    <Row className="selector">
                        <Input s={5} type='select' label="Select Currency" value={this.state.from_element.id} onChange={this.onFromInputChange}>
                            {
                                this.props.exchanges.map((currency) =>
                                    <option key={currency.id} value={currency.id}>{currency.name}</option>
                                )
                            }
                        </Input>
                        <Col s={2} className="arrow">
                            <Icon medium>arrow_forward</Icon>
                        </Col>
                        <Input s={5} type='select' label="Select Currency" value={this.state.to_element.id} onChange={this.onToInputChange}>
                            {
                                this.props.exchanges.map((currency) =>
                                    <option key={currency.id} value={currency.id}>{currency.name}</option>
                                )
                            }
                        </Input>
                    </Row>
                </div>
            </div>
        );
    }

}

const mapDispatchToProps = (dispatch) => {
    return {
        startSetExchange: () => dispatch(startSetExchange())
    };
};

const mapStateToProps = (state, props) => {
    console.log('mapping state');
    return {
        exchanges: state.exchange,
        from_element: state.exchange.find((element) => element.id == 'BTC'),
        to_element: state.exchange.find((element) => element.id == 'USD')
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExchangePage);

