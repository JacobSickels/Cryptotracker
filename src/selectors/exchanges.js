export default (exchanges) => {

    const filteredExchanges = exchanges.slice(0);

    filteredExchanges.sort((a, b) => {
        return a.name > b.name;
    });


    return filteredExchanges;
};
