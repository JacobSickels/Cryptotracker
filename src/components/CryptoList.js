import React from 'react';
import { connect } from 'react-redux';
import selectCryptos from '../selectors/cryptos';

import CryptoListItem from './CryptoListItem';

export const CryptoList = (props) => (
    <div> 
        <h2>CryptoList</h2> 
        Bitcoin:
        <CryptoListItem crypto={props.cryptos.bitcoin} />
        Litecoin:
        <CryptoListItem crypto={props.cryptos.litecoin} /> 
        Ethereum:
        <CryptoListItem crypto={props.cryptos.ethereum} />
    </div>
);

const mapStateToProps = (state) => {
    return {
        cryptos: selectCryptos(state.cryptos, state.filters)
    };
};

export default connect(mapStateToProps)(CryptoList);