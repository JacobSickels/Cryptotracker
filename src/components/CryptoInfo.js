import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { Row, Col } from 'react-materialize';

import { selectCryptos, getMaxMinData } from '../selectors/cryptos';

export class CryptoInfo extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            maxDifference: () => {
                const crypto = props.data;
    
                let maxDifference = 0;
                let minElement = crypto[0].amount;
                
                for(var i = 1; i < crypto.length; i++) {
                    if(crypto[i].amount - minElement > maxDifference) {
                        maxDifference = crypto[i].amount - minElement;
                    }
                    if(crypto[i] < minElement) {
                        minElement = crypto[i].amount;
                    }
                }
                return maxDifference.toFixed(2);
            }
        }
    }

    render() {
        return (
            <div className="chart-info">
                <div className="container">
                    <Row> 
                        <Col s={12} m={4} l={4}>
                            <h1> Max Increase </h1>
                            {this.props.maxDifference}
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Buy</h1> 
                            <em>{moment(this.props.minElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.minElement.amount}</p>
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Sell</h1>
                            <em>{moment(this.props.maxElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.maxElement.amount}</p>
                        </Col>
                    </Row>
                </div>
            </div>
        );
    }

}

const mapStateToProps = (state, props) => {
    const data = getMaxMinData(state.cryptos, state.filters, props.name.toLowerCase());
    return { ...data }
};

export default connect(mapStateToProps)(CryptoInfo);