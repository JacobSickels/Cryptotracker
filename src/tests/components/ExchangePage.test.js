import React from 'react';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { ExchangePage } from '../../components/ExchangePage';
import exchanges from '../fixtures/exchanges'

let store;
let startSetExchange;
let wrapper;

const toExchange = {
    "exchange_rate": "20.14",
    "id": "ARS",
    "name": "Argentine Peso",
    "symbol": "$"
};

const fromExchange = {
    "exchange_rate": "69.50",
    "id": "AFN",
    "name": "Afghan Afghani",
    "symbol": "Ø‹"
};

beforeEach(() => {
    const createMockStore = configureMockStore([thunk]);
    store = createMockStore();
    startSetExchange = jest.fn();

    wrapper = shallow(
        <ExchangePage store={store} 
        startSetExchange={startSetExchange} 
        to_element={toExchange} 
        from_element={fromExchange}
        exchanges={exchanges} 
        />);
});

test('should render exchange page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Use Case #4.1: Should handle amount change', () => {
    const value = "12345.64";
    wrapper.find('#amount-input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('amount')).toEqual(value);
});

test('Use Case $4.2: Should handle base currency change', () => {
    const value = toExchange.id;
    wrapper.find('#base-currency').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('from_element')).toEqual(toExchange);
});

test('Use Case #4.3: Should handle conversion currency change', () => {
    const value = toExchange.id;
    wrapper.find('#conversion-currency').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('to_element')).toEqual(toExchange);
});

test('should handle onAmountChange when amount is NaN', () => {
    const value = "abc123";
    wrapper.find('#amount-input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('amount')).toEqual(1);
});

