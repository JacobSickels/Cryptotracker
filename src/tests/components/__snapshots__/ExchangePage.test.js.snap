// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExchangePage correctly 1`] = `
<div
  className="page-container"
>
  <div
    className="exchange-display"
  >
    <div
      className="container"
    >
      <h3>
         
        1
         
        Afghan Afghani
      </h3>
      <h3>
         
        0.29
         
        Argentine Peso
      </h3>
    </div>
  </div>
  <div
    className="container exchange"
  >
    <Row
      className="selector"
    >
      <Col
        s={5}
      >
        <input
          id="amount-input"
          onChange={[Function]}
          placeholder="Amount"
          type="text"
          value={1}
        />
      </Col>
      <Col
        className="arrow"
        s={2}
      >
        <Icon
          medium={true}
        >
          arrow_forward
        </Icon>
      </Col>
      <div
        s={5}
      >
        <h3>
          0.29
        </h3>
      </div>
    </Row>
    <Row
      className="selector"
    >
      <Input
        checked={false}
        id="base-currency"
        onChange={[Function]}
        s={5}
        type="select"
        value="AFN"
      >
        <option
          key="AED"
          value="AED"
        >
          United Arab Emirates Dirham
        </option>
        <option
          key="AFN"
          value="AFN"
        >
          Afghan Afghani
        </option>
        <option
          key="ALL"
          value="ALL"
        >
          Albanian Lek
        </option>
        <option
          key="AMD"
          value="AMD"
        >
          Armenian Dram
        </option>
        <option
          key="ANG"
          value="ANG"
        >
          Netherlands Antillean Gulden
        </option>
        <option
          key="AOA"
          value="AOA"
        >
          Angolan Kwanza
        </option>
        <option
          key="ARS"
          value="ARS"
        >
          Argentine Peso
        </option>
      </Input>
      <Col
        className="arrow"
        s={2}
      >
        <Icon
          medium={true}
        >
          arrow_forward
        </Icon>
      </Col>
      <Input
        checked={false}
        id="conversion-currency"
        onChange={[Function]}
        s={5}
        type="select"
        value="ARS"
      >
        <option
          key="AED"
          value="AED"
        >
          United Arab Emirates Dirham
        </option>
        <option
          key="AFN"
          value="AFN"
        >
          Afghan Afghani
        </option>
        <option
          key="ALL"
          value="ALL"
        >
          Albanian Lek
        </option>
        <option
          key="AMD"
          value="AMD"
        >
          Armenian Dram
        </option>
        <option
          key="ANG"
          value="ANG"
        >
          Netherlands Antillean Gulden
        </option>
        <option
          key="AOA"
          value="AOA"
        >
          Angolan Kwanza
        </option>
        <option
          key="ARS"
          value="ARS"
        >
          Argentine Peso
        </option>
      </Input>
    </Row>
  </div>
</div>
`;
